code/modules/mob/living/simple_animal/hostile/nanoswarm.dm

```diff
@@ -0,0 +1,166 @@
+	else
+		icon_state = "nanoswarm[round(upgradeLevel)]_dead"
+
+/mob/living/simple_animal/hostile/nanoswarm/proc/updateDamage()
+	melee_damage_lower = min(50,(initial(melee_damage_lower) + (powerStored/1000)*upgradeLevel)/2)
+	melee_damage_upper = min(100,initial(melee_damage_upper) + (powerStored/1000)*upgradeLevel)
+
+/mob/living/simple_animal/hostile/nanoswarm/Life()
+	..()
+	if(health <= 0)
+		return
+	//process movement
+	var/dirchoice = pick(NORTH,SOUTH,EAST,WEST,NORTHEAST,NORTHWEST,SOUTHEAST,SOUTHWEST)
+	var/turf/myTurf = get_turf(src)
+	var/turf/t = get_step(src,dirchoice)
+	var/obj/machinery/M = locate(/obj/machinery) in t
+	var/obj/item/weapon/reagent_containers/food/F = locate(/obj/item/weapon/reagent_containers/food) in t
+	var/obj/item/stack/sheet/S = locate(/obj/item/stack/sheet) in t
+
+	var/obj/structure/cable/C = locate(/obj/structure/cable) in myTurf
+	var/obj/machinery/power/smes/SM = locate(/obj/machinery/power/smes) in t
+	var/obj/machinery/power/apc/A = locate(/obj/machinery/power/apc) in t
+
+	var/mob/living/carbon/human/H = locate(/mob/living/carbon/human) in t
+
+	if(t)
+		if(C)
+			var/drain = C.powernet.avail
+			C.powernet.load += drain
+			powerStored += drain
+			updateDamage()
+
+		if(SM)
+			if(SM.charge)
+				var/drain = SM.charge
+				SM.charge -= drain
+				powerStored += drain
+				updateDamage()
+
+		if(A)
+			if(A.cell && A.cell.charge)
+				var/drain = A.cell.charge
+				A.cell.charge -= drain
+				powerStored += drain
+				updateDamage()
+
+		if(M)
+			M.emagged = 1
+
+		if(S)
+			upgradeLevel = min(5,upgradeLevel + 0.1)
+			qdel(S)
+			update_icon()
+
+		if(F)
+			qdel(F)
+			var/mob/living/simple_animal/hostile/nanoswarm/NS = new /mob/living/simple_animal/hostile/nanoswarm(t)
+			var/shared = powerStored/(max(0.1,2-upgradeLevel))
+			powerStored -= shared
+			NS.powerStored += shared
+
+		if(H)
+			var/ableToInject = 0
+			if(H.head && H.head.flags & THICKMATERIAL)
+				ableToInject++
+			if(H.wear_suit && H.wear_suit.flags & THICKMATERIAL)
+				ableToInject++
+			if(ableToInject < 2)
+				H << "<span class='danger'>You feel a thousand tiny little pricks sizzle across your skin</span>"
+				H.AddDisease(new /datum/disease/transformation/robot/evil)
+
+		if(istype(t,/turf/simulated/floor) || istype(t,/turf/space))
+			if(istype(t,/turf/simulated/floor/nano))
+				Goto(get_step_away(src,t,7),0,0)
+			else
+				t.ChangeTurf(floorConvert)
+				Goto(t,0,0)
+
+		if(istype(t,/turf/simulated/wall))
+			t.ChangeTurf(wallConvert)
+
+
+
+//nanoswarm hives
+/obj/structure/nanohive
+	name = "nanohive"
+	desc = "The surface crawls with a sickly, silver sheen"
+	icon_state = "nanohive"
+	var/maxBots = 5
+	var/curBots = 0
+	var/interval = 30
+	var/health = 100
+
+/obj/structure/nanohive/proc/spawnNano()
+	if(health <= 0)
+		return
+	else
+		for(var/turf/T in oview(1))
+			if(curBots < maxBots)
+				curBots++
+				new /mob/living/simple_animal/hostile/nanoswarm(T)
+		spawn(interval)
+			spawnNano()

```

Comment:
They take extra damage from EMP devices.

Liberal use of grenades and various EMPs is how they were designed to be taken down, not melee'd one by one.